lib/vector-core/src/fanout.rs

```diff
@@ -42,13 +37,66 @@ impl Fanout {
 
         let fanout = Self {
             sinks: vec![],
-            i: 0,
             control_channel: control_rx,
         };
 
         (fanout, control_tx)
     }
 
+    // TODO make a real error
+    pub async fn send_all(&mut self, events: Vec<Event>) -> Result<(), ()> {
+        self.process_control_messages().await;
+
+        if self.sinks.is_empty() {
+            // nothing, intentionally
+        } else if self.sinks.len() == 1 {
+            if let Some(sink) = &mut self.sinks[0].1 {
+                for event in events {
+                    sink.feed(event).await?;
+                }
+                sink.flush().await?;
+            }
+        } else {
+            let mut faulty_sinks = Vec::with_capacity(0);
+
+            {
+                let mut jobs = FuturesUnordered::new();
+                let mut clone_army: Vec<Vec<Event>> = Vec::with_capacity(self.sinks.len());
+                for _ in 0..(self.sinks.len() - 1) {
+                    clone_army.push(events.clone());
+                }
+                clone_army.push(events);
+
+                for (id, ms) in self.sinks.iter_mut() {
+                    if let Some(sink) = ms.as_mut() {
+                        let events: Vec<Event> = clone_army.pop().unwrap();
+                        jobs.push(async move {
+                            for event in events {
+                                sink.feed(event).await.map_err(|e| (id.clone(), e))?
+                            }
+                            sink.flush().await.map_err(|e| (id.clone(), e))
+                        });
+                    }
+                }
+
+                while let Some(res) = jobs.next().await {
+                    match res {
+                        Ok(()) => continue,
+                        Err((id, ())) => {
+                            faulty_sinks.push(id);
+                        }
+                    }
```

Comment:
_nitpick:_ `if let` is a little more concise here:
```suggestion
                    if let Err((id, ())) = res {
                        faulty_sinks.push(id);
                    }
```